packages/cli/src/decorators/controller.registry.ts

```diff
@@ -0,0 +1,136 @@
+import { Container, Service } from 'typedi';
+import { Router } from 'express';
+import type { Application, Request, Response, RequestHandler } from 'express';
+import { rateLimit as expressRateLimit } from 'express-rate-limit';
+
+import { AuthService } from '@/auth/auth.service';
+import config from '@/config';
+import { UnauthenticatedError } from '@/errors/response-errors/unauthenticated.error';
+import { inProduction, RESPONSE_ERROR_MESSAGES } from '@/constants';
+import type { BooleanLicenseFeature } from '@/Interfaces';
+import { License } from '@/License';
+import type { AuthenticatedRequest } from '@/requests';
+import { send } from '@/ResponseHelper'; // TODO: move `ResponseHelper.send` to this file
+import { userHasScope } from '@/permissions/checkAccess';
+
+import type {
+	AccessScope,
+	Controller,
+	ControllerMetadata,
+	HandlerName,
+	RateLimit,
+	RouteMetadata,
+} from './types';
+
+const registry = new Map<Function, ControllerMetadata>();
+
+export const getControllerMetadata = (controllerClass: Function) => {
+	let metadata = registry.get(controllerClass);
+	if (!metadata) {
+		metadata = {
+			basePath: '/',
+			middlewares: [],
+			routes: new Map(),
+		};
+		registry.set(controllerClass, metadata);
+	}
+	return metadata;
+};
+
+export const getRouteMetadata = (controllerClass: Function, handlerName: HandlerName) => {
+	const metadata = getControllerMetadata(controllerClass);
+	let route = metadata.routes.get(handlerName);
+	if (!route) {
+		route = {} as RouteMetadata;
+		metadata.routes.set(handlerName, route);
+	}
+	return route;
+};
+
+@Service()
+export class ControllerRegistry {
+	constructor(
+		private readonly license: License,
+		private readonly authService: AuthService,
+	) {}
+
+	activate(app: Application) {
+		for (const controllerClass of registry.keys()) {
+			this.activateController(app, controllerClass);
+		}
+	}
+
+	private activateController(app: Application, controllerClass: Function) {
+		const metadata = registry.get(controllerClass)!;
+
+		const router = Router({ mergeParams: true });
+		const prefix = `/${[config.getEnv('endpoints.rest'), metadata.basePath].join('/')}`
```

Comment:
nit: IMHO a simple string interpolation would be clearer

```suggestion
		const prefix = `/${config.getEnv('endpoints.rest')}/${metadata.basePath}`
```